README.md
<p align="center">
    <img src="https://user-images.githubusercontent.com/20530235/124830986-a5f70200-df48-11eb-8800-22eaed9b21ba.png?sanitize=true">
</p>
<p align="center">
    <img alt="Travis (.com)" src="https://img.shields.io/travis/com/fsjorgeluis/password-generator">
    <img alt="Lines of code" src="https://img.shields.io/tokei/lines/github/fsjorgeluis/password-generator">
    <img alt="npm bundle size" src="https://img.shields.io/bundlephobia/minzip/ts-password-generator">
    <img alt="GitHub package.json dependency version (dev dep on branch)" src="https://img.shields.io/github/package-json/dependency-version/fsjorgeluis/password-generator/dev/typescript">
    <img alt="GitHub package.json dependency version (dev dep on branch)" src="https://img.shields.io/github/package-json/dependency-version/fsjorgeluis/password-generator/dev/jest">
    <img alt="GitHub" src="https://img.shields.io/github/license/fsjorgeluis/password-generator">
    <a href="https://twitter.com/intent/follow?screen_name=fsjorgeluis">
        <img alt="Twitter Follow" src="https://img.shields.io/twitter/follow/fsjorgeluis?logo=twitter&logoColor=blue&style=social">
    </a>
</p>

# Password Generator Module

A little module to generate password, use `fromCharCode()` static method to convert a random "ASCII - Unicode" code to a readable character.

```
String.fromCharCode()
```

Feel free to use it, if you like it! :D

## Ready to use

Just `npm install ts-password-generator`, then

```
import { generator } from 'ts-password-generator';

const password: string = generator({ haveNumbers: true })
```

or

```
const tsPasswordGenerator = require("ts-password-generator");

const tsPass = tsPasswordGenerator;

tsPass.generator({});
```

and you are ready to go.

The `generator` method receive following optional params:

```
{
    charsQty: number,
    isUppercase: boolean,
    haveNumbers: boolean,
    haveString: boolean,
    haveSymbols: boolean
}
```

and returns a brand new password string.

### Accepted params by generator method

Any of these can be passed into the options object for `generator` method.

| Name        | Description                                                                         | Default Value |
| ----------- | ----------------------------------------------------------------------------------- | ------------- |
| charsQty    | Integer, optional param, password length (chars quantity).                          | 10            |
| isUppercase | Boolean, optional param, add uppercase chars to the generated password.             | false         |
| haveNumbers | Boolean, optional param, add numbers to the generated password.                     | false         |
| haveString  | Boolean, optional param, add char to the generated password.                        | true          |
| haveSymbols | Boolean, optional param, add symbols (special characters) to the generated password | false         |

_NOTE: by default, if you don't pass any parameter to the `generator` method, the generated password will be in lowercase._

### TODO

- Add maybe more fuctionality.

### DONE

- Make this little script an npm package just for fun, and educative purposes.
- Added more documentation.


CHANGELOG.md
# Password Generator Changelog

## v1.0.4

_2021-11-08_

- Created haveStrings option
- Validated values for GeneratorOptions for make it optionals (with default values)
- Changed GeneratorOptios name to GeneratorOptions
- If you pass all have.\* values as false the generator function will return a '' string

_Note_: To use the `generator` function with default parameters without passing arguments and passing a `{}` as first argument, the `generator` function apply the default values listed in the [README.md](https://github.com/fsjorgeluis/password-generator/blob/main/README.md) default values table.

## v1.0.3

_2021-07-14_

- Added CHANGELOG

## v1.0.2

_2021-07-07_

- Added more info to readme

## v1.0.1

_2021-07-07_

- Added info, logo and badges to readme

## v1.0.0

_2021-07-07_

- Initial stable release and upload to npmjs as testing


test/index.spec.ts
import { jest } from "@jest/globals";
import { generator } from '../src/';

beforeEach(() => {
    jest.spyOn(global.Math, 'floor').mockReturnValue(23);
});

afterEach(() => {
    jest.spyOn(global.Math, 'floor').mockRestore();
});

test("should return a string of xxxxxxxxx", () => {
    expect(generator({
        charsQty: 10,
        isUppercase: false,
        haveNumbers: false,
        haveSymbols: false
    })).toBe("xxxxxxxxxx");
});

test/generator.spec.ts
import { jest } from "@jest/globals";
// import { lowToHigh } from '../src/generator';
import { generator } from '../src/';

beforeEach(() => {
    jest.spyOn(global.Math, 'floor').mockReturnValue(23);
});

afterEach(() => {
    jest.spyOn(global.Math, 'floor').mockRestore();
});

test("should return a string of xxxxxxxxx", () => {
    expect(generator({
        charsQty: 10,
        isUppercase: false,
        haveNumbers: false,
        haveSymbols: false
    })).toBe("xxxxxxxxxx");
});

// test("should return a array of numbers", () => {
//     expect(lowToHigh(1, 5)).toStrictEqual([1, 2, 3, 4, 5]);
// });

src/generator.ts
// Type definition
type GeneratorOptios = {
    charsQty: number,
    isUppercase: boolean,
    haveNumbers: boolean,
    haveSymbols: boolean
};

/**
 * Prepare an array o numbers based on ASCII table
 * @param low minimum value
 * @param high maximum value
 * @returns array of numbers
 */
const lowToHigh = (low: number, high: number): number[] => {
    const array: number[] = [];
    for (let i = low; i <= high; i++) {
        array.push(i);
    }
    return array;
};

const lowercaseCharactersCode = lowToHigh(97, 122);
const uppercaseCharactersCode = lowToHigh(65, 90);
const numberCharactersCode = lowToHigh(48, 57);
const symbolsCharactersCode = lowToHigh(33, 47)
    .concat(lowToHigh(58, 64))
    .concat(lowToHigh(123, 126))
    .concat(lowToHigh(91, 96));

/**
 * Get named params (object) and loop through it and apply conditionals
 * @param charsQty      string length
 * @param isUppercase   boolean param should include uppercase?
 * @param haveNumbers   boolean param should include numbers?
 * @param haveSymbols   boolean param should include special chars?
 * @returns string with password ready characteristics
 */
export const generator = ({
    charsQty = 10,
    isUppercase = false,
    haveNumbers = false,
    haveSymbols = false
}: GeneratorOptios): string => {
    let charactersCode: number[] = lowercaseCharactersCode;

    if (isUppercase) charactersCode = charactersCode.concat(uppercaseCharactersCode);
    if (haveNumbers) charactersCode = charactersCode.concat(numberCharactersCode);
    if (haveSymbols) charactersCode = charactersCode.concat(symbolsCharactersCode);

    const password: string[] = [];
    for (let i = 0; i < charsQty; i++) {
        const char = Math.floor(Math.random() * charactersCode.length);
        password.push(String.fromCharCode(charactersCode[char]));
    }
    return password.join('');
};

const execute = generator({
    charsQty: 10,
    isUppercase: true,
    haveNumbers: true,
    haveSymbols: true
});

console.log(execute);

src/lib/index.ts
// Type definition
type GeneratorOptions = {
	charsQty?: number;
	isUppercase?: boolean;
	haveString?: boolean;
	haveNumbers?: boolean;
	haveSymbols?: boolean;
};

/**
 * Prepare an array or numbers based on ASCII table
 * @param low minimum value
 * @param high maximum value
 * @returns array of numbers
 */
const lowToHigh = (low: number, high: number): number[] => {
	const array: number[] = [];
	for (let i = low; i <= high; i++) {
		array.push(i);
	}

	return array;
};

const lowercaseCharactersCode = lowToHigh(97, 122);
const uppercaseCharactersCode = lowToHigh(65, 90);
const numberCharactersCode = lowToHigh(48, 57);
const symbolsCharactersCode = lowToHigh(33, 47)
	.concat(lowToHigh(58, 64))
	.concat(lowToHigh(123, 126))
	.concat(lowToHigh(91, 96));

/**
 * Get named params (object) and loop through it and apply conditionals
 * @param charsQty      string length
 * @param isUppercase   boolean param should include uppercase?
 * @param haveNumbers   boolean param should include numbers?
 * @param haveSymbols   boolean param should include special chars?
 * @returns string with password ready characteristics
 */
export const generator = (
	{
		charsQty = 10,
		isUppercase = false,
		haveString = true,
		haveNumbers = false,
		haveSymbols = false,
	}: GeneratorOptions = {
		charsQty: 10,
		isUppercase: false,
		haveString: true,
		haveNumbers: false,
		haveSymbols: false,
	}
): string => {
	let charactersCode: number[] = [];

	if (haveString)
		charactersCode = charactersCode.concat(lowercaseCharactersCode);
	if (isUppercase)
		charactersCode = charactersCode.concat(uppercaseCharactersCode);
	if (haveNumbers) charactersCode = charactersCode.concat(numberCharactersCode);
	if (haveSymbols)
		charactersCode = charactersCode.concat(symbolsCharactersCode);

	const password: string[] = [];
	if (charactersCode.length === 0) {
		return '';
	}
	for (let i = 0; i < charsQty; i++) {
		const char = Math.floor(Math.random() * charactersCode.length);
		password.push(String.fromCharCode(charactersCode[char]));
	}
	return password.join('');
};
